internal/pkg/agent/cmd/inspect.go

```diff
@@ -5,383 +5,444 @@
+	vars, err := waitForVariables(ctx, l, cfg, timeout)
+	if err != nil {
+		return nil, fmt.Errorf("failed to gather variables: %w", err)
 	}
-	for _, iface := range arr {
-		input, ok := iface.(map[string]interface{})
-		if !ok {
-			continue
-		}
-		t, ok := input["type"]
-		if !ok {
-			continue
+
+	// Render the inputs using the discovered inputs.
+	inputs, ok := transpiler.Lookup(ast, "inputs")
+	if ok {
+		renderedInputs, err := transpiler.RenderInputs(inputs, vars)
+		if err != nil {
+			return nil, fmt.Errorf("rendering inputs failed: %w", err)
 		}
-		if t.(string) == "fleet-server" {
-			return input
+		err = transpiler.Insert(ast, renderedInputs, "inputs")
+		if err != nil {
+			return nil, fmt.Errorf("inserting rendered inputs failed: %w", err)
 		}
 	}
-	return nil
+	m, err = ast.Map()
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert ast to map[string]interface{}: %w", err)
+	}
+	return m, nil
 }
 
-type inmemRouter struct {
-	programs map[string][]program.Program
+type varsWait struct {
+	vars []*transpiler.Vars
+	err  error
 }
 
-func (r *inmemRouter) Routes() *sorted.Set {
-	return nil
-}
+func waitForVariables(ctx context.Context, l *logger.Logger, cfg *config.Config, wait time.Duration) ([]*transpiler.Vars, error) {
+	var cancel context.CancelFunc
 
-func (r *inmemRouter) Route(_ context.Context, _ string, grpProg map[pipeline.RoutingKey][]program.Program) error {
-	r.programs = grpProg
-	return nil
-}
+	composable, err := composable.New(l, cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create composable controller: %w", err)
+	}
 
-func (r *inmemRouter) Shutdown() {}
+	hasTimeout := false
+	if wait > time.Duration(0) {
+		hasTimeout = true
+		ctx, cancel = context.WithTimeout(ctx, wait)
+	} else {
+		ctx, cancel = context.WithCancel(ctx)
+	}
+	defer cancel()
 
-type waitForCompose struct {
-	controller composable.Controller
-	done       chan bool
-}
+	resCh := make(chan varsWait)
+	go func() {
+		var err error
+		var vars []*transpiler.Vars
+		for {
+			select {
+			case <-ctx.Done():
+				if err == nil {
+					err = ctx.Err()
+				}
+				if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+					err = nil
+				}
+				resCh <- varsWait{vars: vars, err: err}
+				return
+			case cErr := <-composable.Errors():
+				err = cErr
+				if err != nil {
+					cancel()
+				}
+			case cVars := <-composable.Watch():
+				vars = cVars
+				if !hasTimeout {
+					cancel()
+				}
+			}
+		}
+	}()
 
-func newWaitForCompose(wrapped composable.Controller) *waitForCompose {
-	return &waitForCompose{
-		controller: wrapped,
-		done:       make(chan bool),
+	errCh := make(chan error)
+	go func() {
+		err := composable.Run(ctx)
+		if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+			err = nil
+		}
+		errCh <- err
+	}()
+
+	err = <-errCh
+	if err != nil {
+		return nil, err
 	}
+	res := <-resCh
+	return res.vars, res.err
```

Comment:
maybe worth refactoring with https://pkg.go.dev/golang.org/x/sync/errgroup

Reply:
I think it's not a bad idea what do you think @blakerouse ?

Reply:
I should really use that more often. Updated to use errgroup.